gpio/src/gpio.rs

```diff
@@ -0,0 +1,943 @@
+                VIRTIO_GPIO_DIRECTION_NONE as u32,
+            ))
+        }
+    }
+
+    fn set_value(&self, gpio: u16, value: u32) -> Result<()> {
+        let state = &self.state[gpio as usize].read().unwrap();
+
+        // Direction change can follow value change, don't fail here for invalid
+        // direction.
+        if let Some(request) = &state.request {
+            request
+                .set_value(gpio as u32, value as i32)
+                .map_err(Error::GpiodFailed)?;
+        }
+
+        Ok(())
+    }
+
+    fn set_irq_type(&self, gpio: u16, value: u16) -> Result<()> {
+        let state = &mut self.state[gpio as usize].write().unwrap();
+        let mut config = LineConfig::new().map_err(Error::GpiodFailed)?;
+
+        match value as u16 {
+            VIRTIO_GPIO_IRQ_TYPE_EDGE_RISING => config.set_edge_detection(Edge::Rising),
+            VIRTIO_GPIO_IRQ_TYPE_EDGE_FALLING => config.set_edge_detection(Edge::Falling),
+            VIRTIO_GPIO_IRQ_TYPE_EDGE_BOTH => config.set_edge_detection(Edge::Both),
+
+            // Drop the buffer.
+            VIRTIO_GPIO_IRQ_TYPE_NONE => {
+                state.buffer = None;
+                return Ok(());
+            }
+
+            // Only edge IRQs are supported for now.
+            _ => return Err(Error::GpioIrqTypeNotSupported(value)),
+        };
+
+        // Allocate the buffer and configure the line for interrupt.
+        if state.request.is_none() {
+            Err(Error::GpioLineRequestNotConfigured)
+        } else {
+            state.buffer = Some(EdgeEventBuffer::new(1).map_err(Error::GpiodFailed)?);
+
+            state
+                .request
+                .as_ref()
+                .unwrap()
+                .reconfigure_lines(&config)
+                .map_err(Error::GpiodFailed)
+        }
+    }
+
+    fn wait_for_interrupt(&self, gpio: u16) -> Result<()> {
+        // While waiting here for the interrupt to occur, it is possible that we receive another
+        // request to disable the interrupt instead, via a call to set_irq_type().
+        //
+        // That request shouldn't wait for this wait_for_interrupt() function to finish, else we
+        // may never proceed with set_irq_type() as the parent wait_for_interrupt() function will
+        // take the lock right after dropping one and hence set_irq_type() will wait for ever.
+        //
+        // To avoid this situation, implement the state.request as an "Arc" instance instead and
+        // take a reference to it without locking the state. With this, while we are waiting for
+        // the interrupt, set_irq_type() can come and set the state.request to None and we will be
+        // able to quit the wait loop.
+        let request = {
+            let state = &self.state[gpio as usize].write().unwrap();
+
+            match &state.request {
+                Some(x) => x.clone(),
+                None => return Err(Error::GpioIrqNotEnabled),
+            }
+        };
```

Comment:
In the above block the line state is locked and the increment count of the LineRequest increased.  The lock is released on line 294...